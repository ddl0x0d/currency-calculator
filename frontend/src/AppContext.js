import React from 'react';

const AppContext = React.createContext({ currencies: [] });

export default AppContext;
