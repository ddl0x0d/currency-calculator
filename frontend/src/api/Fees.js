import axios from 'axios';
import { extractMessages } from './errorService';

const api = '/api/fees';

export const get = async () => {
  try {
    const response = await axios.get(api);
    return response.data.data;
  } catch (e) {
    throw extractMessages(e);
  }
};

export const add = async fee => {
  try {
    const response = await axios.post(api, fee);
    return response.data.data;
  } catch (e) {
    throw extractMessages(e);
  }
};

export const remove = async id => {
  try {
    return await axios.delete(`${api}/${id}`);
  } catch (e) {
    if (e.response && e.response.status === 404) {
      console.warn(`Fee with ID "${id}" not found`);
    } else {
      throw extractMessages(e);
    }
  }
};

export default { get, add, remove };
