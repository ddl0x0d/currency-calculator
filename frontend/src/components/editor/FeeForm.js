import AmountInput from '../common/AmountInput';
import AppContext from '../../AppContext';
import Button from 'react-bootstrap/Button';
import Col from 'react-bootstrap/Col';
import CurrencySelector from '../common/CurrencySelector';
import Fees from '../../api/Fees';
import Form from 'react-bootstrap/Form';
import PropTypes from 'prop-types';
import React, { useContext, useState } from 'react';
import Row from 'react-bootstrap/Row';

const propTypes = {
  addFee: PropTypes.func.isRequired
};

const FeeForm = props => {
  const [from, setFrom] = useState('');
  const [to, setTo] = useState('');
  const [amount, setAmount] = useState(0);
  const { showErrors } = useContext(AppContext);

  const addFee = () => {
    Fees.add({ from, to, amount })
      .then(fee => {
        props.addFee(fee);
        showErrors([]);
      })
      .catch(errors => {
        showErrors(errors);
      });
  };

  return (
    <Row className="justify-content-center">
      <Col xs={3} md={2} xl={1}>
        <CurrencySelector
          id="from"
          label="From"
          value={from}
          handleChange={value => setFrom(value)}
        />
      </Col>
      <Col xs={3} md={2} xl={1}>
        <CurrencySelector
          id="to"
          label="To"
          value={to}
          handleChange={value => setTo(value)}
        />
      </Col>
      <Col xs={3} md={2} xl={1}>
        <AmountInput
          id="amount"
          label="Fee"
          value={amount}
          step={0.01}
          handleChange={value => setAmount(value)}
        />
      </Col>
      <Col xs={3} md={2} xl={1}>
        <Form.Label className="mb-0">&nbsp;</Form.Label>
        <Form.Group>
          <Button
            id="add"
            block="true"
            disabled={!(from && to && amount) || from === to}
            size="sm"
            onClick={() => addFee()}
          >
            Add
          </Button>
        </Form.Group>
      </Col>
    </Row>
  );
};

FeeForm.propTypes = propTypes;

export default FeeForm;
