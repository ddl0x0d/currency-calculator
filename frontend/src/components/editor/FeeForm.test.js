import AppContext from '../../AppContext';
import FeeForm from './FeeForm';
import Fees from '../../api/Fees';
import { mount, shallow } from 'enzyme';
import React from 'react';

jest.mock('../../api/Fees');

describe('<FeeForm />', () => {
  const addFee = jest.fn().mockName('addFee');
  const feeForm = () => <FeeForm addFee={addFee} />;

  it('renders correctly', () => {
    const wrapper = mount(feeForm());
    expect(wrapper).toMatchSnapshot();
  });

  describe('"Add" button', () => {
    const wrapper = shallow(feeForm());
    const addDisabled = () => wrapper.find('#add').prop('disabled');
    const fillForm = (from, to, amount) => {
      wrapper.find('#from').prop('handleChange')(from);
      wrapper.find('#to').prop('handleChange')(to);
      wrapper.find('#amount').prop('handleChange')(amount);
    };

    it('is disabled by default', () => {
      expect(addDisabled()).toBe(true);
    });

    it('is disabled when "From" and "To" are equal', () => {
      fillForm('EUR', 'EUR', 0.01);
      expect(addDisabled()).toBe(true);
    });

    it('is enabled', () => {
      fillForm('EUR', 'GBP', 0.01);
      expect(addDisabled()).toBe(false);
    });
  });

  describe('add fee', () => {
    beforeEach(() => {
      jest.clearAllMocks();
    });

    const showErrors = jest.fn().mockName('showErrors');
    const wrapper = mount(
      <AppContext.Provider value={{ currencies: [], showErrors }}>
        {feeForm()}
      </AppContext.Provider>
    );

    const change = (selector, value) => {
      const element = wrapper.find(selector);
      element.simulate('change', { target: { value } });
    };

    change('select#from', 'EUR');
    change('select#to', 'GBP');
    change('input#amount', 0.01);
    const feeRequest = { from: 'EUR', to: 'GBP', amount: 0.01 };

    it('succeeds', done => {
      const fee = { id: '123', ...feeRequest };
      Fees.add.mockName('Fees.add').mockResolvedValue(fee);

      wrapper.find('button#add').simulate('click');

      setTimeout(() => {
        expect(Fees.add).toBeCalledWith(feeRequest);
        expect(addFee).toBeCalledWith(fee);
        expect(showErrors).toBeCalledWith([]);
        done();
      });
    });

    it('fails', done => {
      const errors = [{ message: 'error' }];
      Fees.add.mockName('Fees.add').mockRejectedValue(errors);

      wrapper.find('button#add').simulate('click');

      setTimeout(() => {
        expect(Fees.add).toBeCalledWith(feeRequest);
        expect(addFee).not.toBeCalled();
        expect(showErrors).toBeCalledWith(errors);
        done();
      });
    });
  });
});
