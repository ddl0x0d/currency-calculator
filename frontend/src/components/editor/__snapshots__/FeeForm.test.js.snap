// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FeeForm /> renders correctly 1`] = `
<FeeForm
  addFee={[MockFunction addFee]}
>
  <Bootstrap(Row)
    className="justify-content-center"
  >
    <Row
      as="div"
      bsPrefix="row"
      className="justify-content-center"
      noGutters={false}
    >
      <div
        className="justify-content-center row"
      >
        <Col
          as="div"
          md={2}
          xl={1}
          xs={3}
        >
          <div
            className="col-xl-1 col-md-2 col-3"
          >
            <CurrencySelector
              handleChange={[Function]}
              id="from"
              label="From"
              value=""
            >
              <FormGroup
                as="div"
              >
                <div
                  className="form-group"
                >
                  <FormLabel
                    className="small mb-0"
                    column={false}
                    srOnly={false}
                  >
                    <label
                      className="small mb-0 form-label"
                    >
                      From
                    </label>
                  </FormLabel>
                  <FormControl
                    as="select"
                    id="from"
                    onChange={[Function]}
                    size="sm"
                    value=""
                  >
                    <select
                      className="form-control form-control-sm"
                      id="from"
                      onChange={[Function]}
                      value=""
                    >
                      <option
                        key=""
                        value=""
                      />
                    </select>
                  </FormControl>
                </div>
              </FormGroup>
            </CurrencySelector>
          </div>
        </Col>
        <Col
          as="div"
          md={2}
          xl={1}
          xs={3}
        >
          <div
            className="col-xl-1 col-md-2 col-3"
          >
            <CurrencySelector
              handleChange={[Function]}
              id="to"
              label="To"
              value=""
            >
              <FormGroup
                as="div"
              >
                <div
                  className="form-group"
                >
                  <FormLabel
                    className="small mb-0"
                    column={false}
                    srOnly={false}
                  >
                    <label
                      className="small mb-0 form-label"
                    >
                      To
                    </label>
                  </FormLabel>
                  <FormControl
                    as="select"
                    id="to"
                    onChange={[Function]}
                    size="sm"
                    value=""
                  >
                    <select
                      className="form-control form-control-sm"
                      id="to"
                      onChange={[Function]}
                      value=""
                    >
                      <option
                        key=""
                        value=""
                      />
                    </select>
                  </FormControl>
                </div>
              </FormGroup>
            </CurrencySelector>
          </div>
        </Col>
        <Col
          as="div"
          md={2}
          xl={1}
          xs={3}
        >
          <div
            className="col-xl-1 col-md-2 col-3"
          >
            <AmountInput
              handleChange={[Function]}
              id="amount"
              label="Fee"
              step={0.01}
              value={0}
            >
              <FormGroup
                as="div"
              >
                <div
                  className="form-group"
                >
                  <FormLabel
                    className="small mb-0"
                    column={false}
                    srOnly={false}
                  >
                    <label
                      className="small mb-0 form-label"
                    >
                      Fee
                    </label>
                  </FormLabel>
                  <FormControl
                    as="input"
                    id="amount"
                    min={0}
                    onChange={[Function]}
                    size="sm"
                    step={0.01}
                    type="number"
                    value={0}
                  >
                    <input
                      className="form-control form-control-sm"
                      id="amount"
                      min={0}
                      onChange={[Function]}
                      step={0.01}
                      type="number"
                      value={0}
                    />
                  </FormControl>
                </div>
              </FormGroup>
            </AmountInput>
          </div>
        </Col>
        <Col
          as="div"
          md={2}
          xl={1}
          xs={3}
        >
          <div
            className="col-xl-1 col-md-2 col-3"
          >
            <FormLabel
              className="mb-0"
              column={false}
              srOnly={false}
            >
              <label
                className="mb-0 form-label"
              >
                Â 
              </label>
            </FormLabel>
            <FormGroup
              as="div"
            >
              <div
                className="form-group"
              >
                <Button
                  active={false}
                  block="true"
                  disabled={true}
                  id="add"
                  onClick={[Function]}
                  size="sm"
                  type="button"
                  variant="primary"
                >
                  <button
                    className="btn btn-primary btn-block btn-sm"
                    disabled={true}
                    id="add"
                    onClick={[Function]}
                    type="button"
                  >
                    Add
                  </button>
                </Button>
              </div>
            </FormGroup>
          </div>
        </Col>
      </div>
    </Row>
  </Bootstrap(Row)>
</FeeForm>
`;
